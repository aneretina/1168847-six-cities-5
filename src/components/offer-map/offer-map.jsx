import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import leaflet from 'leaflet';
import 'leaflet/dist/leaflet.css';
import {ZOOM, Icon, ID_MAP_CONTAINER, CitiesCoordinates} from "../../const.js";

class OfferMap extends PureComponent {
  constructor(props) {
    super(props);
    this._map = null;

    this.defaultIcon = leaflet.icon({
      iconUrl: Icon.URL,
      iconSize: Icon.SIZE
    });

    this.activeIcon = leaflet.icon({
      iconUrl: Icon.ACTIVE_URL,
      iconSize: Icon.SIZE
    });

    this._pins = new Map();
  }

  componentDidMount() {
    const {offers, city} = this.props;

    this._map = leaflet.map(ID_MAP_CONTAINER, {
      center: CitiesCoordinates[city.toUpperCase()],
      zoom: ZOOM,
      zoomControl: false,
      marker: true
    });

    leaflet
  .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
    attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a>contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
  })
  .addTo(this._map);

    this._addPins(offers);
  }

  componentDidUpdate(prevProps) {
    const {offers, city, offerId} = this.props;

    if (city !== prevProps.city) {
      this._removePins();
      this._setView();
    }

    this._addPins(offers);

    if (offerId !== prevProps.offerId) {
      this._deactivateMapPin(prevProps.offerId);
      this._activateMapPin(offerId);
    }
  }

  _addPins() {
    const {offerId, offers} = this.props;
    offers.forEach((offer) => {
      const pin = leaflet
    .marker(offer.location, {icon: this.defaultIcon})
    .addTo(this._map);
      this._pins.set(offer.id, pin);
    });
    this._activateMapPin(offerId);
  }

  _activateMapPin(offerId) {
    const marker = this._pins.get(offerId);
    if (marker) {
      marker.setIcon(this.activeIcon);
    }
  }

  _deactivateMapPin(offerId) {
    const pin = this._pins.get(offerId);
    if (pin) {
      pin.setIcon(this.defaultIcon);
    }
  }

  _setView() {
    this._map.setView(CitiesCoordinates[this.props.city.toUpperCase()], ZOOM);
  }

  _removePins() {
    this._pins.forEach((pin) => {
      pin.removeFrom(this._map);
    });
    this._pins.clear();
  }

  render() {
    const {className} = this.props;
    return (
      <section className={`map ${className}`} id="map"></section>
    );
  }
}

OfferMap.propTypes = {
  offers: PropTypes.array.isRequired,
  className: PropTypes.string.isRequired,
  city: PropTypes.string.isRequired,
  offerId: PropTypes.number.isRequired
};

const mapStateToProps = (state) => ({
  offers: state.currentCityOffers,
  offerId: state.activeOfferId,
  city: state.city
});

export {OfferMap};
export default connect(mapStateToProps)(OfferMap);
