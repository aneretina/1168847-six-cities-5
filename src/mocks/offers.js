export const offers = [
  {
    id: 15,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.38333,
        longitude: 4.9,
      }
    },
    location: {
      latitude: 52.3909553943508,
      longitude: 4.85309666406198,
    },
    pictures: [`img/apartment-01.jpg`, `img/apartment-01.jpg`, `img/apartment-01.jpg`, `img/apartment-01.jpg`, `img/apartment-01.jpg`, `img/apartment-01.jpg`],
    previewImage: `img/apartment-01.jpg`,
    isPremium: true,
    isFavorite: true,
    price: 180,
    title: `Wood and stone place`,
    type: `room`,
    rating: 3,
    description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
    bedroomsCount: 2,
    guestsLimit: 2,
    goods: [`Wi-Fi`, `Heating`, `Kitchen`, `Parking`, `Flowers`, `Dishwasher`, `Towels`, `TV`],
    host: {
      avatar: `img/avatar-max.jpg`,
      name: `Max`,
      isPro: false,
    }
  },

  {
    id: 2,
    city: {
      name: `Paris`,
      location: {
        latitude: 48.864716,
        longitude: 2.349014,
      }
    },
    location: {
      latitude: 48.87456,
      longitude: 2.413,
    },
    pictures: [`img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`],
    previewImage: `img/apartment-02.jpg`,
    isPremium: true,
    isFavorite: true,
    price: 80,
    title: `Beautiful & luxurious apartment in Paris`,
    type: `Apartment`,
    rating: 4,
    description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
    bedroomsCount: 1,
    guestsLimit: 2,
    goods: [`Wi-Fi`, `Heating`, `Kitchen`, `Parking`, `Flowers`, `Dishwasher`, `Towels`, `TV`],
    host: {
      avatar: `img/avatar-maggy.jpg`,
      name: `Maggy`,
      isPro: true,
    }
  },

  {
    id: 9,
    city: {
      name: `Cologne`,
      location: {
        latitude: 50.935173,
        longitude: 6.953101,
      }
    },
    location: {
      latitude: 50.9682,
      longitude: 6.874,
    },
    pictures: [`img/apartment-03.jpg`, `img/apartment-03.jpg`, `img/apartment-03.jpg`, `img/apartment-03.jpg`, `img/apartment-03.jpg`, `img/apartment-03.jpg`],
    previewImage: `img/apartment-03.jpg`,
    isPremium: false,
    isFavorite: false,
    price: 220,
    title: `Beautiful & luxurious apartment in Cologne`,
    type: `Apartment`,
    rating: 1,
    description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
    bedroomsCount: 4,
    guestsLimit: 5,
    goods: [`Wi-Fi`, `Heating`, `Kitchen`, `Parking`, `Flowers`, `Dishwasher`, `Towels`, `TV`],
    host: {
      avatar: `img/avatar-bill.jpg`,
      name: `Bill`,
      isPro: true,
    }
  },
];
