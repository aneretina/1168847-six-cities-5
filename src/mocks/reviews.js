export default [
  {
    id: 1,
    name: `Jake`,
    avatar: [`https://placeimg.com/640/480/people/101`,
      `https://placeimg.com/640/480/people/102`,
      `https://placeimg.com/640/480/people/103`,
      `https://placeimg.com/640/480/people/104`,
    ],
    rating: 3,
    text: `We would like to thank all the staff, who were very friendly during our stay.`,
    date: `10.10.2019`,
  },

  {
    id: 2,
    name: `Mary`,
    avatar: [`https://placeimg.com/640/480/people/101`,
      `https://placeimg.com/640/480/people/102`,
      `https://placeimg.com/640/480/people/103`,
      `https://placeimg.com/640/480/people/104`,
    ],
    rating: 2,
    text: `The hotel had everything we needed and it was in an excellent location, being in a lovely area with lots of places to enjoy ourselves.`,
    date: `10.09.2019`,
  },

  {
    id: 3,
    name: `Oliver`,
    avatar: [`https://placeimg.com/640/480/people/101`,
      `https://placeimg.com/640/480/people/102`,
      `https://placeimg.com/640/480/people/103`,
      `https://placeimg.com/640/480/people/104`,
    ],
    rating: 5,
    text: `Iâ€™ve been many times and I will come again and again.`,
    date: `10.09.2020`,
  },


];
