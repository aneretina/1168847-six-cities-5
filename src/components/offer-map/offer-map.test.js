import React from "react";
import renderer from 'react-test-renderer';
import {OfferMap} from "./offer-map";
import {offers} from "../../mocks/offers";


it(`Map render correctly`, () => {
  const tree = renderer
    .create(
        <OfferMap
          offers={offers}
          className={`cities__map`}
          offerId={1}
          cityCoords={[52.43212, 2.657]}
          zoom={12}
          mainOffer={offers[1]}
        />,
        {createNodeMock: () => {
          return document.createElement(`div`);
        }}
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
