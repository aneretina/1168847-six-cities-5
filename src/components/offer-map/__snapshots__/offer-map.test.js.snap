// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map render correctly 1`] = `
<section
  className="map cities__map"
  id="map"
/>
`;
